File: gpt-4/template_context_full/parallel_construct_vector_length_clause.cpp
#include "acc_testsuite.h"
#include <cmath>

#ifndef T1
/*T1:parallel construct vector_length clause,V:2.0-2.7*/
int test1(){
    int err = 0;
    srand(SEED);

    int vector_length = 512;
    int *a, *b, *c;
    int n = vector_length * 4;

    a = new int[n];
    b = new int[n];
    c = new int[n];

    for (int i = 0; i < n; i++) {
        a[i] = rand() % 100;
        b[i] = rand() % 100;
    }

    #pragma acc parallel copyout(c[0:n]) copyin(a[0:n], b[0:n]) vector_length(vector_length)
    {
        #pragma acc loop worker
        for(int i = 0; i < n; i++){
            c[i] = a[i] * b[i];
        }
    }

    for (int i=0; i < n; i++) {
        if (c[i] != a[i] * b[i]) {
            err = 1;
        }
    }
    
    delete[] a;
    delete[] b;
    delete[] c;

    return err;
}
#endif

int main(){
    int failcode = 0;
    int failed;
#ifndef T1
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test1();
    }
    if (failed != 0){
        failcode = failcode + (1 << 0);
    }
#endif
    return failcode;
}

File: gpt-4-turbo/stage2/parallel_construct_vector_length_clause.f90
! This code snippet is meant to be part of a larger testsuite, and it requires
! an OpenACC implementation and external test framework support to run.
! Assume acc_testsuite.Fh handles test framework initialization and shutdown procedures.

!T1:parallel construct vector_length clause,V:2.7-2.3
      LOGICAL FUNCTION test1()
        USE OPENACC
        IMPLICIT NONE
        INCLUDE "acc_testsuite.Fh"
        
        INTEGER, PARAMETER :: N = 1024
        INTEGER :: errors = 0
        REAL, DIMENSION(N) :: a
        INTEGER :: i

        ! Initialize array
        a = 1.0

        ! Use parallel construct with vector_length clause
        !$acc parallel loop vector_length(128)
        DO i = 1, N
          a(i) = a(i) + 1.0
        END DO
        !$acc end parallel

        ! Check results
        DO i = 1, N
          IF (a(i) /= 2.0) THEN
            errors = errors + 1
          END IF
        END DO

        IF (errors .eq. 0) THEN
          test1 = .FALSE.
        ELSE
          test1 = .TRUE.
        END IF
      END

      PROGRAM main
        IMPLICIT NONE
        INTEGER :: failcode, testrun
        LOGICAL :: failed
        INCLUDE "acc_testsuite.Fh"
        EXTERNAL test1
        LOGICAL :: test1
        failed = .FALSE.
        failcode = 0
        DO testrun = 1, NUM_TEST_CALLS
          failed = failed .or. test1()
        END DO
        IF (failed) THEN
          failcode = failcode + 2 ** 0
          failed = .FALSE.
        END IF
        CALL EXIT (failcode)
      END PROGRAM